arch/Xtensa/XtensaDisassembler.c

```diff
@@ -0,0 +1,302 @@
+{
+	CS_ASSERT(CONCAT(isUInt, 5)(Imm) && "Invalid immediate");
+	MCOperand_CreateImm0(Inst, (Imm));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeImm1_16Operand(MCInst *Inst, uint64_t Imm,
+					 int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 4)(Imm) && "Invalid immediate");
+	MCOperand_CreateImm0(Inst, (Imm + 1));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeShimm1_31Operand(MCInst *Inst, uint64_t Imm,
+					   int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 5)(Imm) && "Invalid immediate");
+	MCOperand_CreateImm0(Inst, (32 - Imm));
+	return MCDisassembler_Success;
+}
+
+static int64_t TableB4const[16] = { -1, 1,  2,	3,  4,	5,  6,	 7,
+				    8,	10, 12, 16, 32, 64, 128, 256 };
+static DecodeStatus decodeB4constOperand(MCInst *Inst, uint64_t Imm,
+					 int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 4)(Imm) && "Invalid immediate");
+
+	MCOperand_CreateImm0(Inst, (TableB4const[Imm]));
+	return MCDisassembler_Success;
+}
+
+static int64_t TableB4constu[16] = { 32768, 65536, 2,  3,  4,  5,  6,	7,
+				     8,	    10,	   12, 16, 32, 64, 128, 256 };
+static DecodeStatus decodeB4constuOperand(MCInst *Inst, uint64_t Imm,
+					  int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 4)(Imm) && "Invalid immediate");
+
+	MCOperand_CreateImm0(Inst, (TableB4constu[Imm]));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeMem8Operand(MCInst *Inst, uint64_t Imm,
+				      int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 12)(Imm) && "Invalid immediate");
+	DecodeARRegisterClass(Inst, Imm & 0xf, Address, Decoder);
+	MCOperand_CreateImm0(Inst, ((Imm >> 4) & 0xff));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeMem16Operand(MCInst *Inst, uint64_t Imm,
+				       int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 12)(Imm) && "Invalid immediate");
+	DecodeARRegisterClass(Inst, Imm & 0xf, Address, Decoder);
+	MCOperand_CreateImm0(Inst, ((Imm >> 3) & 0x1fe));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeMem32Operand(MCInst *Inst, uint64_t Imm,
+				       int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 12)(Imm) && "Invalid immediate");
+	DecodeARRegisterClass(Inst, Imm & 0xf, Address, Decoder);
+	MCOperand_CreateImm0(Inst, ((Imm >> 2) & 0x3fc));
+	return MCDisassembler_Success;
+}
+
+/// Read three bytes from the ArrayRef and return 24 bit data
+static DecodeStatus readInstruction24(MCInst *MI, uint64_t *SizeOut,
+				      const uint8_t *Bytes,
+				      const unsigned BytesSize, uint32_t *Insn)
+{
+	// We want to read exactly 3 Bytes of data.
+	if (BytesSize < 3) {
+		*SizeOut = 0;
+		return MCDisassembler_Fail;
+	}
+
+	if (MODE_IS_BIG_ENDIAN(MI->csh->mode))
+		*Insn = (Bytes[2]) | (Bytes[1] << 8) |
+			((uint32_t)Bytes[0] << 16);
+	else
+		*Insn = (Bytes[2] << 16) | (Bytes[1] << 8) |
+			((uint32_t)Bytes[0]);
```

Comment:
Can you move this please to `utils.c::readBytes24()`? There are already others of this kind.